### ISSUE TITLE
RLlib Previous actions when action 0 is used

### ISSUE THREAD

**Issue by dejangrubisic** [ray-project/ray#32927]
### What happened + What you expected to happen

I am building the custom model and want to use previous actions as part of the state. When I register that I want prev_actions in input_dict, input_dict['prev_actions '] will be filled up with zeros, so when I apply action 0 in the beginning, input_dict['prev_actions '] will not change, so I won't know if I applied action 0 or not. Is there some other way to figure out the length of prev_actions?

### Versions / Dependencies

ray 2.2.0

### Reproduction script

```
class TorchCustomModel(TorchModelV2, nn.Module):
    """Example of a PyTorch custom model that just delegates to a fc-net."""

    def __init__(self, obs_space, action_space, num_outputs, model_config, name):

        .....

        # Add prev-action/reward nodes to input to LSTM.
        if self.use_n_prev_actions:
            self.num_outputs += self.use_n_prev_actions * self.action_dim

        if self.use_n_prev_actions:
            self.view_requirements['prev_actions'] = ViewRequirement(
                'actions',
                space=self.action_space,
                shift="-{}:-1".format(self.use_n_prev_actions),
            )

    def forward(self, input_dict, state, seq_lens):
        print(input_dict.keys(), f"Prev_actions: {input_dict['prev_actions'] if  'prev_actions' in input_dict else ''}")


```

Output when running the script:
```
dict_keys(['obs', 'prev_actions', 'obs_flat']) Prev_actions: tensor([[0, 0, 0, 0, 0]])
Applying action 0, equivalent command-line arguments: '-add-discriminators'

dict_keys(['obs', 'prev_actions', 'obs_flat']) Prev_actions: tensor([[0, 0, 0, 0, 0]])
Applying action 0, equivalent command-line arguments: '-add-discriminators'

dict_keys(['obs', 'prev_actions', 'obs_flat']) Prev_actions: tensor([[0, 0, 0, 0, 0]])
Applying action 1, equivalent command-line arguments: '-adce'

dict_keys(['obs', 'prev_actions', 'obs_flat']) Prev_actions: tensor([[0, 0, 0, 0, 1]])
```

### Issue Severity

Low: It annoys or frustrates me.

**Comment by ArturNiederfahrenhorst** [comment]
Hi @dejangrubisic ,

Can you provide a full reproductions script? 
Some more info: 

- The Model itself does not keep track of the state. Doing a forward pass that results in some action manually won't change future prev_actions. 
- The size of `prev_actions` depends solely on the number of timesteps in the SampleBatch.
- The first `prev_actions` of each trajectory should be zero. Keep in mind that the new gymnasium API has what was once "soft_horizon" in RLlib - soft resets.

**Comment by ArturNiederfahrenhorst** [comment]
If the reproduction script could contain a `print()` with the output and a `print()` with the desired output next to it, it might also become more clear to me where we diverge from your expected behaviour.

**Comment by dejangrubisic** [comment]
Thanks for the answer, before writing full code, let me try to explain visually the problem.
![image](https://user-images.githubusercontent.com/40608743/224449376-87ecbd50-faa8-4cb4-bdfb-8fea5b97384b.png)

Before applying any action forward() is called with prev_actions [0, 0, 0, 0, 0], 

Then we apply the first action - say action 0 - > forward is called with [0, 0, 0, 0, **0**] (the last number is action we just applied == 0)

Question: How do we know if we applied action 0, once, twice, three, four or five times? 
To answer this I had to implement additional logic, is there a better way?

**Comment by ArturNiederfahrenhorst** [comment]
@dejangrubisic Thank you, I understand your question now. And thanks for opening an issue.
In the very beginning, we generally don't know. With respect to your example, the model "thinks" that it has applied `action=0` 4 four times. If the model uses all these actions as inputs, then there is no elegant way around this. You could, in theory, use `action=-inf` or `action=123` - you'll have to find _some_ value to represent these missing pieces. Usually, zero is a good choice because model inputs are commonly normalized.
Feel free to reopen this issue if you have further questions or input on this.
